File: apps/apple/TopsyTurvy/TopsyTurvyToolchainTests/TestCallbackCapture.swift
import TopsyTurvyToolchain

/// Holds outputs from callbacks, used for testing purposes.
///
/// A `@convention(c)` function cannot capture instance state, so the callback targets below are top-level
/// functions that read and write the static state of this type instead.
final class TestCallbackCapture {
    /// Holds the lines of output captured by `captureOutputLine`.
    static var outputLines: [String] = []

    /// The line returned by `provideInputLine`, or `nil` to simulate no input available.
    static var inputLineResult: String?
}

/// The `output_line` callback target registered with `topsyturvy_tc_session_create`.
///
/// - Parameters:
///   - context: The context pointer passed to `topsyturvy_tc_session_create`.
///   - utf8: The UTF-8 string to capture.
///   - suppressNewline: A value indicating whether the string should be appended to the last line of output (if non-zero) or added as a new line (if zero).
func captureOutputLine(context: UnsafeMutableRawPointer?, utf8: UnsafePointer<UInt8>?, suppressNewline: UInt8) {
    guard let utf8 else {
        return
    }

    let message = String(cString: utf8)
    if suppressNewline != 0, !TestCallbackCapture.outputLines.isEmpty {
        TestCallbackCapture.outputLines[TestCallbackCapture.outputLines.count - 1] += message
    } else {
        TestCallbackCapture.outputLines.append(message)
    }
}

/// The `resolve_import` callback target registered with `topsyturvy_tc_session_create`.
///
/// This always returns `nil`, since these tests do not import.
///
/// - Parameters:
///   - context: The context pointer passed to `topsyturvy_tc_session_create`.
///   - filename: The UTF-8 string containing the filename to resolve.
///
/// - Returns: A pointer to the imported source text.
func resolveImportStub(context: UnsafeMutableRawPointer?, filename: UnsafePointer<UInt8>?) -> UnsafeMutablePointer<UInt8>? {
    nil
}

/// The `input_line` callback target registered with `topsyturvy_tc_session_create`.
///
/// Returns `TestCallbackCapture.inputLineResult` as a caller-owned buffer, or `nil` if none was configured.
///
/// - Parameters:
///   - context: The context pointer passed to `topsyturvy_tc_session_create`.
///
/// - Returns: A pointer to the configured input line, or `nil`.
func provideInputLine(context: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<UInt8>? {
    guard let inputLineResult = TestCallbackCapture.inputLineResult else {
        return nil
    }

    guard let copy = strdup(inputLineResult) else {
        return nil
    }

    return UnsafeMutableRawPointer(copy).assumingMemoryBound(to: UInt8.self)
}

File: operetta/BWHazel.TopsyTurvy.Tests/Embedded/TestCallbackCapture.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Text;

namespace BWHazel.TopsyTurvy.Tests.Embedded;

/// <summary>
/// Provides <see cref="UnmanagedCallersOnlyAttribute"/> callback targets for testing <c>tt_session_create</c>.
/// </summary>
/// <remarks>
/// Callback function pointers cannot capture instance state, so captured output is held in a static list
/// and the import-resolution/input-line results are each held in a static field.  Tests using this class
/// must call <see cref="Reset"/> before each session under test.
/// </remarks>
internal static unsafe class TestCallbackCapture
{
    private static readonly List<string> outputLines = [];
    private static string? importResolutionResult;
    private static string? inputLineResult;

    /// <summary>
    /// Gets the output lines captured by <see cref="OnOutputLine"/> since the last <see cref="Reset"/>.
    /// </summary>
    internal static IReadOnlyList<string> OutputLines => outputLines;

    /// <summary>
    /// Clears captured output, the import-resolution result and the input-line result.
    /// </summary>
    internal static void Reset()
    {
        outputLines.Clear();
        importResolutionResult = null;
        inputLineResult = null;
    }

    /// <summary>
    /// Sets the string returned by <see cref="OnResolveImport"/>, or <c>null</c> to simulate an unresolvable import.
    /// </summary>
    /// <param name="result">The import source text to return, or <c>null</c>.</param>
    internal static void SetImportResolutionResult(string? result) => importResolutionResult = result;

    /// <summary>
    /// Sets the string returned by <see cref="OnInputLine"/>, or <c>null</c> to simulate no input available.
    /// </summary>
    /// <param name="result">The input line to return, or <c>null</c>.</param>
    internal static void SetInputLineResult(string? result) => inputLineResult = result;

    /// <summary>
    /// Captures a single output line, appending to the previous line instead of starting a new one when suppressed.
    /// </summary>
    /// <param name="context">The opaque context pointer, unused by this test target.</param>
    /// <param name="utf8Message">A null-terminated, UTF-8 encoded pointer to the message.</param>
    /// <param name="suppressNewline">Non-zero when the message continues the previous output line.</param>
    [UnmanagedCallersOnly]
    internal static void OnOutputLine(nint context, byte* utf8Message, byte suppressNewline)
    {
        string message = Marshal.PtrToStringUTF8((nint)utf8Message) ?? string.Empty;
        if (suppressNewline != 0 && outputLines.Count > 0)
        {
            outputLines[^1] += message;
        }
        else
        {
            outputLines.Add(message);
        }
    }

    /// <summary>
    /// Returns the configured import-resolution result as a caller-owned UTF-8 buffer.
    /// </summary>
    /// <param name="context">The opaque context pointer, unused by this test target.</param>
    /// <param name="filenameUtf8">The requested import filename, unused by this test target.</param>
    /// <returns>A null-terminated, UTF-8 encoded buffer, or a null pointer if no result was configured.</returns>
    [UnmanagedCallersOnly]
    internal static byte* OnResolveImport(nint context, byte* filenameUtf8)
    {
        if (importResolutionResult is null)
        {
            return null;
        }

        int byteCount = Encoding.UTF8.GetByteCount(importResolutionResult);
        nint buffer = Marshal.AllocHGlobal(byteCount + 1);
        Span<byte> destination = new((void*)buffer, byteCount + 1);
        Encoding.UTF8.GetBytes(importResolutionResult, destination);
        destination[byteCount] = 0;
        return (byte*)buffer;
    }

    /// <summary>
    /// Returns the configured input-line result as a caller-owned UTF-8 buffer.
    /// </summary>
    /// <param name="context">The opaque context pointer, unused by this test target.</param>
    /// <returns>A null-terminated, UTF-8 encoded buffer, or a null pointer if no result was configured.</returns>
    [UnmanagedCallersOnly]
    internal static byte* OnInputLine(nint context)
    {
        if (inputLineResult is null)
        {
            return null;
        }

        int byteCount = Encoding.UTF8.GetByteCount(inputLineResult);
        nint buffer = Marshal.AllocHGlobal(byteCount + 1);
        Span<byte> destination = new((void*)buffer, byteCount + 1);
        Encoding.UTF8.GetBytes(inputLineResult, destination);
        destination[byteCount] = 0;
        return (byte*)buffer;
    }
}
